Update offset section in the encoder article

Daehoon-Sung opened the pull request:
This PR is to update offset section in the encoder article.

Daehoon-Sung committed:
Change indexme file

Daehoon-Sung committed:
Update index.md

elsevers committed:
Edit finding offset and step 1

elsevers committed:
Edit finding the offset section

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -89,22 +89,31 @@ double task_get_theta_m(void)
+4. Repeat step 3 at several speeds.
+5. For each offset value, average the d-axis voltage values for different speeds and select the offset with the lowest average d-axis voltage.
+6. Plot the d-axis voltage with the selected offset against the different rotor speeds. The d-axis voltage should be close to zero for all speeds, if the offset is tuned correctly.
+7. In-case there is an error in the offset value, a significant speed-dependent voltage will appear on the d-axis voltage. In this case, the user may have to re-measure the encoder offset.
```
## Confusion in steps 3 - 7

Steps 3 through 7 are a bit inconsistent (and I think my edits to step 3 may have made this worse... sorry...). 

It sounds like you want the user to sweep `enc_theta_m_offset` over a fixed range at each speed, save the voltage data for each value of `enc_theta_m_offset`, at each speed, and then at the end determine which `enc_theta_m_offset` minimizes everything. However, that isn't really captured in the text of the current steps. 

This also raises a few questions: 
1. how do you know what range of `enc_theta_m_offset` values to use? Maybe at the first speed you do this at you identify the `enc_theta_m_offset` value that produces a zero crossing and then use +-5 of that value and this is the set of `enc_theta_m_offset` values you use at all speeds.
2. At the end, when finding the `enc_theta_m_offset`that minimizes everything, it seems like you really should normalize this by speed, since the error voltage you'll see in the d-axis is proportional to speed: <img width="586" alt="image" src="https://github.com/user-attachments/assets/9ddced61-ebaf-48a6-be27-87b2f0cb58c1" />.
3. Also, while at high speed the signal-to-noise ratio is higher, we should acknowledge that there is likely more torque on the shaft due to bearing friction and windage and so a user should be aware of this and may need to compensate for it (i..e., $i_q \neq 0$)
4. To illustrate this / make it easy, we should provide an example (with data and necessary MATLAB scripts to interpret it all), like we do with [current sensor calibration](https://docs.amdc.dev/getting-started/control-with-amdc/current-sensor-cal/index.html#calibration-method)

## Path forward

### 1. the steps

Okay, having said all of that, you have a choice here:
1. Proceed along the path you are on (explain this procedure of sampling at several speeds, taking a weighted average)
2. As an alternate to collecting an array of $v_d$ values for several `enc_theta_m_offset` values at each speed, you could instead have the user find the `enc_theta_m_offset` value that minimizes $v_d$ at each speed, and then take the average of `enc_theta_m_offset` at the end. This this would be much easier to explain and you wouldn't have to have to think about normalizing your results by speed.
3. Change the procedure to only consider testing at a single speed and briefly tell them that they could further refine their accuracy by repeating this at other speeds and averaging the results.

**I recommend** either option 2 or option 3.

### 2. Add supporting example material

Regardless of the path you go down, let's have you add some example scripts/data and results.

I'd really like to see a plot like this at the end (but only showing 1 machine):
![image](https://github.com/user-attachments/assets/f7e2fab5-bdf9-42b9-b80e-36d93e8b0462)

Daehoon-Sung replied:
Thank you for your feedback Professor @elsevers. I revised the step2 following your second feedback. Plus, I added the plot.

elsevers commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -89,22 +89,31 @@ double task_get_theta_m(void)
-1. In-case there is an error in the offset value, a significant speed dependent voltage will appear on the d-axis voltage. In this case, the user may have to re-measure the encoder offset.
+#### Step 2: Determine precise offset
 
+Friction and cogging torque in the motor decrease the accuracy of the estimate in Step 1. The precise offset can be found by fine-tuning the `enc_theta_m_offset` from Step 1 while using closed-loop control to rotate the shaft at significant speed. This can be done as follows:
```
Let's provide the d-axis voltage expression and have a sentence explaining how the technique works with respect to this (i.e., no loading so $i_q = 0$, we're commanding $i_d=0$, and therefore any voltage observed in the d axis must be due to encoder offset error).

elsevers reviewed (changes requested):
Thank you @Daehoon-Sung. This is very nice. I made some edits and have a couple requests -- see below.

elsevers committed:
Merge branch 'add-enc-anirudh' into user/Daehoon-Sung/update-offset-report

elsevers committed:
Add torque characteristic and update step 1 instructions.

elsevers committed:
Merge branch 'add-enc-anirudh' into user/Daehoon-Sung/update-offset-report

elsevers committed:
Add clarifying comment to torque characteristic

Daehoon-Sung committed:
Update index.md

Apply feedback from Professor's comment (option 2) to revise the steps to find offset

Daehoon-Sung committed:
Update the encoder offset plot

Daehoon-Sung committed:
Update index.md

Update the image

Daehoon-Sung committed:
Update index.md

Add the formula

Daehoon-Sung commented:
Hello, Dr @noguchi-takahiro. Could you review this PR when you have time? Thank you!

noguchi-takahiro commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -108,22 +108,56 @@ double task_get_theta_m(void)
+
+The approximate encoder offset can be found by taking advantage of the motor having the torque characteristic shown on the right. This depicts shaft torque as the shaft is rotated counter-clockwise and corresponds to [the image at the start of the section](#rotor-position); positive torque is in the counter-clockwise direction.
+
+The following simple procedure can be used without any feedback control:
```
Here when you are saying the "without any feedback control", does this mean that we need to convey  "without any speed feedback control"? This is because you still need to regulate the current using feedback control if you inject a current command on the d-axis? @elsevers @Daehoon-Sung

elsevers replied:
Thanks @noguchi-takahiro. I think it is okay as-is because we refer to using a DC supply to control current into the phases.

noguchi-takahiro replied:
Thank you, but I think the article tells us that we have two approaches to achieve it:

![image](https://github.com/user-attachments/assets/5d94569c-f70a-482f-811e-7b71647ed3dc)

This means that users might use the AMDC's signal injection feature still, so current regulation with feedback control is required.

Also, although this is an optional change, in practice it is not necessary to use the signal injection module. But instead, we can simply set the input angle `theta_m_enc` to 0 and command the DC d-axis current as in normal current regulation, i.e., not use signal injection module.

noguchi-takahiro commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -108,22 +108,56 @@ double task_get_theta_m(void)
+
+#### Step 2: Determine precise offset
+
+Friction and cogging torque in the motor decrease the accuracy of the estimate in Step 1. The precise offset can be found by fine-tuning the `enc_theta_m_offset` from Step 1 while using closed-loop control to rotate the shaft at significant speed. The correct offset is determined by observing $\hat{\theta}_e$ using the following $v_d$ equation. 
```
When you say "significant speed" here, I wonder this means that like high-speed? We know the BP4 has spined at 3000 RPM at that time, which  I believe it was not significant speed. I understand that we should do this at the higher speed, but in practice, we sometimes cannot do so. I suggest something like this:

```suggestion
Friction and cogging torque in the motor decrease the accuracy of the estimate in Step 1. The precise offset can be found by fine-tuning the `enc_theta_m_offset` from Step 1 while using closed-loop control to rotate the shaft at  at the highest possible speed. The correct offset is determined by observing $\hat{\theta}_e$ using the following $v_d$ equation. 
```

noguchi-takahiro commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -108,22 +108,56 @@ double task_get_theta_m(void)
+Friction and cogging torque in the motor decrease the accuracy of the estimate in Step 1. The precise offset can be found by fine-tuning the `enc_theta_m_offset` from Step 1 while using closed-loop control to rotate the shaft at significant speed. The correct offset is determined by observing $\hat{\theta}_e$ using the following $v_d$ equation. 
+
+$$
+v_d = (R + pL) i_d - \hat{\omega}\_e L i_q - \hat{\omega}\_e \lambda\_{\mathrm{pm}} \sin(\tilde{\theta}\_e)
```
This is optional, but do you think the $L$ here could be $L_d$ or $L_q$? I understand that they are same value for the SPM, i.e., non-salient machine.

Daehoon-Sung replied:
Thank you. I revised this.

noguchi-takahiro commented on source/getting-started/control-with-amdc/encoder-fb/index.md:
```diff
@@ -108,22 +108,56 @@ double task_get_theta_m(void)
-1. Plot the d-axis voltage against the rotor speed
-1. The d-axis voltage should be close to zero for all speeds, if the offset is tuned correctly
-1. In-case there is an error in the offset value, a significant speed dependent voltage will appear on the d-axis voltage. In this case, the user may have to re-measure the encoder offset.
+1. Configure the AMDC for closed-loop speed and DQ current control and configure the operating environment to allow for quick edits to `enc_theta_m_offset` and for measuring the d-axis voltage commanded by the current regulator. Consider [adding a custom command](/getting-started/tutorials/vsi/index.md#command-template-c-code) and using [logging](/getting-started/user-guide/logging/index.md) to a ...
```
In these steps, looks like these 1) Step, 2) step, and 3) part are used inconsistently. Please consider to use them consistently.

Daehoon-Sung replied:
Hello @noguchi-takahiro. I don't understand this. There are largely two steps and each step has each own procedure.

#### Step 1 
Step 1 explanation
1. 
2. 
3. ...

#### Step 2 
Step 2 explanation
1. 
2. 
3. ...

May I ask what makes you think it is inconsistent?

noguchi-takahiro replied:
We use Step 1 here, i.e., starting with capital:
![image](https://github.com/user-attachments/assets/11b750a2-2b14-4f54-bba3-36dcfbe21cd7)

We use step 3 here, i.e., starting with lower case:
![image](https://github.com/user-attachments/assets/8ab4c6e1-3f7b-4caa-8109-6557b82837ea)

Then we use part 4:

![image](https://github.com/user-attachments/assets/9ab10f74-5fff-4918-a668-3c05ebe72b42)

I wanted to recommend to use them all Step to be consistent.

noguchi-takahiro reviewed (changes requested):
Thank you @Daehoon-Sung. This looks good. I left some minor suggestions for you. Could you consider to address them?

Daehoon-Sung committed:
Update source/getting-started/control-with-amdc/encoder-fb/index.md

Co-authored-by: Takahiro <114006024+noguchi-takahiro@users.noreply.github.com>

Daehoon-Sung committed:
Update index.md

elsevers reviewed (changes requested):
Hi @Daehoon-Sung,

## 1. Equations

> <img width="300" alt="image" src="https://github.com/user-attachments/assets/ce858693-e0a2-4dcc-bc2e-0dfe60a7eb86" />

1. Can you fix the issue with these subscripts not rending correctly?
2. Can use the angles as defined in the images above ($\theta_{\rm off}$, $\theta_{\rm enc}$, $\theta_{\rm m}$):

> <img width="200" alt="image" src="https://github.com/user-attachments/assets/15b3b882-3358-49c5-864b-ce23adf06e11" />
> <img width="300" alt="image" src="https://github.com/user-attachments/assets/690c876f-a728-40b6-b43b-cbfb103a483a" />

## 2. Variable names
My edits of the code yesterday changed the variables. Can you go through your steps and update the variables names to match what is used in the code above?

> <img width="500" alt="image" src="https://github.com/user-attachments/assets/fc43508f-13d8-4795-acb1-ce083a75aa4a" />


## 3. Image not rendering

This image is not appearing when I build the page:
> <img width="700" alt="image" src="https://github.com/user-attachments/assets/12a9ba6f-b0e8-4fe4-9e5d-dac370aef798" />

## 4. Do we have any better data?

This data shows some pretty significant errors at low speeds in the calibrated waveform. Do we have any other data that is more convincing? Or, you could use this data but plot only the data points above 1000 RPM.

> <img width="300" alt="image" src="https://github.com/user-attachments/assets/71e164ad-aecf-48b5-9f32-a8f75a0f58de" />

Daehoon-Sung committed:
Update index.md

Daehoon-Sung committed:
Update index.md

noguchi-takahiro committed:
Fix rendering of equation

noguchi-takahiro committed:
Fix rendering issue of image

Daehoon-Sung committed:
Update the encoder image

Daehoon-Sung committed:
Merge branch 'user/Daehoon-Sung/update-offset-report' of https://github.com/Severson-Group/docs.amdc.dev into user/Daehoon-Sung/update-offset-report

Daehoon-Sung committed:
Update index.md

Update the variable names

noguchi-takahiro reviewed (approved):
Thank you @Daehoon-Sung. Looks good to me.

@elsevers could you review this PR?

Daehoon-Sung commented:
Hello Professor @elsevers. Could you review this?

noguchi-takahiro reviewed (changes requested):
Per 11/26 meeting, 

- In the `Determine precise offset` section, show this equation that everyone knows:
<img width="631" height="304" alt="Image" src="https://github.com/user-attachments/assets/6bdfc40f-e547-4d20-8b6f-e0f4fe7c640b" />

- Use the following image and explain it logically, showing more general case:
<img width="560" height="273" alt="Image" src="https://github.com/user-attachments/assets/f2ff7614-d341-46e1-b534-8f3fa8c61bf4" />

- Show this:
<img width="528" height="243" alt="Image" src="https://github.com/user-attachments/assets/bd1c142c-c5a4-44fe-8cba-4aff57cef382" />

- Show this, where currents are 0:
<img width="780" height="215" alt="Image" src="https://github.com/user-attachments/assets/27333ec3-65d1-4bbb-b356-3bfeba7e3cae" />


- The idea it that we should clarify the meaning of angle $\theta_m$ and continue to read through the entire article without confusion, might need to define $\theta_m$ and $\theta_e$.
- Think if we really need to define $\tilde{\theta_e}$
- Define $\omega_e$, that you can include this in the figure as well


See complete notes here:
[Encoder Article Vgamma calc.pdf](https://github.com/user-attachments/files/23777483/Encoder.Article.Vgamma.calc.pdf)

Daehoon-Sung committed:
Update the figure

Daehoon-Sung committed:
Clarify terminology for mechanical angle in documentation

Daehoon-Sung committed:
Enhance encoder offset determination section

Added detailed explanation and equations for determining encoder offset using closed-loop control and voltage measurements.

Daehoon-Sung committed:
Clarify voltage equation with angular velocity note

Added clarification about electrical angular velocity in the voltage equation.

Daehoon-Sung committed:
Clarify voltage vector description in index.md

Daehoon-Sung committed:
Fix voltage vector equation formatting

Corrected the formatting of the voltage vector equation in complex vector form.

Daehoon-Sung committed:
Refine description of encoder offset estimation process

Clarified the conditions for estimating encoder offset by refining language regarding the alignment of the gamma-delta and d-q frames.

Daehoon-Sung committed:
Fix minor grammatical error in encoder offset section

Daehoon-Sung committed:
Fix formatting in encoder offset determination section

Daehoon-Sung committed:
Revise current command notation in encoder equations

Updated notation for current commands in the equation for v_gamma.

Daehoon-Sung commented:
Thank you very much for [this comment](https://github.com/Severson-Group/docs.amdc.dev/pull/139#pullrequestreview-3512138559) @noguchi-takahiro. I have carefully implemented all the feedback. Could you review this?

noguchi-takahiro committed:
Fix rendering issues for images

noguchi-takahiro reviewed (commented):
Thank you @Daehoon-Sung. This looks great. I have a couple small requests here:

- Here is a rendered version of the article [Encoder Feedback - AMDC Platform.pdf](https://github.com/user-attachments/files/23894807/Encoder.Feedback.-.AMDC.Platform.pdf), and I am noting that these `below` and `above` actually should be `right`? What do you think?

<img width="545" height="291" alt="Image" src="https://github.com/user-attachments/assets/18008c70-6a44-4f06-910f-8877c8a9c577" />

- I wonder if we defined $\theta_\mathrm{e}$ somewhere?

Daehoon-Sung committed:
Clarify angle calculation and encoder offset procedure

Updated explanations regarding the calculation of the motor's angle and the determination of the encoder offset. Improved clarity on the relationship between electrical angle and voltage vector.

Daehoon-Sung commented:
Hello Dr @noguchi-takahiro. Thank you for the feedback. I revised it and also clarified the electrical angle and mechanical angle. 
Could you review this? Thank you.

noguchi-takahiro reviewed (approved):
Thank you @Daehoon-Sung. Looks good to me.

@elsevers could you review this PR?

elsevers reviewed (changes requested):
@Daehoon-Sung, nice improvements. 

I have made suggested edits for you in #139. 

Please edit/merge these in and re-request my review. I think it is good to go after you merge these in.

elsevers committed:
Revise encoder offset section contents (#146)

* Edit encoder offset precise section

* Improve typesetting and clarify vd

Daehoon-Sung commented:
Hello Professor @elsevers. I have merged the PR #146 to here. Could you review the article? Thank you.

elsevers reviewed (approved):
Nice work. Thank you @Daehoon-Sung
